### DISCUSSION TITLE
do whisper need preprocessing, enhancing, or denoising audio before transcribing?

### DISCUSSION THREAD

**Discussion by FLabbaf97** [openai/whisper#2125] [Q&A] [answered]
Hello.
I want to transcribe short voices with Whisper large v3. It works fine for some of the files, But in some cases, the transcription is not very accurate. I realized my audio files have a high noise level. (average -30 db). some of this noise is from the background, some are from the microphone. Do you think enhancing audio quality and denoising will help? does whisper basicaly need denoising?

The problem is that I couldn't find a single method that enhances the quality of all of my files. every method works for some files, but reduce the speech quality on other files. I will be happy if you also know any method that wisely enhance speech quality. 

Thank you for your support.

**Comment by itaipee** [comment]
Generally speaking , in speech recognition , de-noising rarely works.   First ,because many methods work for some noises but not for others . Second ,  the de-noising modify the spectral representation of the audio .   Dumping the noise , usually means that the certain frequency bands are distorted ,  and even when for human , the speech sounds much more clear , for the ASR , the spectral representation is less resemble the pattern of the expected audio.  

An exception for that, ,is when the ASR is trained on de-noised audio , so it becomes "familiar" with spectrum representations of de-noised audio. 
It can be considered as some sort of augmentation.
However, I doubted Whisper was trained that way.

**Reply by EtienneAb3d** [reply, marked as answer]
@itaipee 
Rather than ordinary de-noiser, you can use neural voice extraction tools, like [Facebook Demucs](https://github.com/facebookresearch/demucs). 
It works very well !
See several processing tools combinations (Demucs, Silence/Noise removal, voice compression, etc) with this experimental code:
https://github.com/EtienneAb3d/WhisperHallu

**Reply by itaipee** [reply]
thanks  @EtienneAb3d    I will certainty look Whisper-Hallu   it looks interesting , and i wonder if it works well on actual audio. 

P.S.   isn't Demucs used for music ?  does it work well for "cleaning" speech  , without any distortion for the speech recognition system?

**Reply by gongouveia** [reply]
@itaipee Yes, demucs is originally for music separation, however in this version it was manipulated for voice enhancement and background noise removal (https://github.com/facebookresearch/denoiser ).
The version i Cited is based on demucsvv2 using and LSTM in and the version above is based on DemucsV4 using a transformer.

**Reply by EtienneAb3d** [reply]
@itaipee 

> P.S. isn't Demucs used for music ? does it work well for "cleaning" speech , without any distortion for the speech recognition system?

Demucs is splitting sound in 4 stems: Voice, Bass, Drums, .. and Others. The last one is very efficiently filled with all being something else than what is properly in the first 3 parts. In my tests, it works very well for the de-noise purpose.

**Reply by FLabbaf97** [reply]
Thank you. Demucs worked very well for denoising my audio in most cases.
However, in some cases, using Demucs diminishes the speech quality. Do you have any suggestion on how to approach this problem? I want to automate everything so I don't have to check the audio quality manually. 
Thanks.

**Reply by gongouveia** [reply]
@FLabbaf97 It is possible to do it. Firts gather some files with clean audio (no noisy), and just trash them with your expected noise distribution.
Now you have a set of clear and noisy files, you can compare them using some Audio enhancement metrics, such as PESQ, STOI, POLQA.
This is technique is more used when you are training an enhancement model, or just tuning it, so you valuate its progress.
If you have already the model and you want to compare the vocals quality live, that requires a different type of setup.

**Reply by KleberNoel** [reply]
Another option is to indeed use speech enhancement. I have successfully tried this approach when diarizatiom fails. I believe there is a speech enhancement model somewhere under speechbrain's hugging face repository.

**Reply by ngcheeyuan** [reply]
DEMUCS uses lots of memory, and it grows with audio length. Does it make sense to split the audio and run DEMUCS on it and recombine it? Is there any downside?

I've tried using Facebook denoiser and spectral gating and both approach decreased  performance and at times distort the actual speech

**Reply by gongouveia** [reply]
@ngcheeyuan Facebook denoisers works great. Yes you can split the audio in smaller chunks and recombine it with no loss of accuracy.

**Reply by ngcheeyuan** [reply]
i've tried https://github.com/facebookresearch/denoiser on my data set, with and without VAD and it generally made WER worse.

I've tried https://github.com/facebookresearch/demucs previously but couldn't continue doing it because some of my audio files are rather long and took up too much memory.  Are you refering to this?

**Reply by gongouveia** [reply]
Hi @ngcheeyuan , 
I am referring to using this tool [facebook denoiser](https://github.com/facebookresearch/denoiser )
I don't have full knowledge of your dataset, it should decrease the WER of noisy dataset. if dataset is not very noisy probably it is just deleting important frequency components that whisper is looking for. 

Do not use that version of Demucs (V3) because it is transformer based, as you say it is expected to increase the resources needed with audio input length. Look for the Demucs (V2) LSTM based bottleneck [[link]](https://github.com/facebookresearch/demucs/tree/v2), the lstm should not eat up your ram.

Hope I have helped you. Feel free to make more question here or in DM.

**Reply by dgoryeo** [reply]
@klebster2 , do you perhaps mean `SpectralMaskEnhancement` model?

**Reply by KleberNoel** [reply]
That's right. I believe it is either that type of model or `SepformerSeparation`. Can't remember exactly though, I tried a couple, and since then lost the code, because this was an edge case.
If I remember correctly, before using the enhancement I was not able to get any output for diarization using pyannote's model, and after using enhancement, there was output, and it was acceptable (what I percieved as speech with low SNR (high noise) was finally diarized).
It might have been either of these (speechbrain lists several more - and as I recall it is quite GPU hungry):

- https://huggingface.co/speechbrain/metricgan-plus-voicebank
- https://huggingface.co/speechbrain/sepformer-wham-enhancement

**Reply by joelai0101** [reply]
I attempted to create noisy and enhanced datasets using the following:

LibriSpeech validation sets with added White Gaussian Noise. The noisy datasets were denoised using [facebook denoiser](https://github.com/facebookresearch/denoiser).
Evaluations were conducted on ESPNet using OpenAI's Whisper large-v2 to calculate the WER at different SNR levels.
However, the WER after denoising turned out to be worse. I am currently trying to use Demucs-v2 for denoising but the dataset processing speed is not fast enough.

**Reply by asccm** [reply]
Hi @joelai0101, did it improve with any of the Demucs versions? Curious to know the outcome ....
